src/ddspdrum/torchmodule.py

```diff
@@ -0,0 +1,410 @@
+
+torch.pi = torch.acos(torch.zeros(1)).item() * 2  # which is 3.1415927410125732
+
+
+class TorchSynthModule(nn.Module):
+    """
+    Base class for synthesis modules, in torch.
+
+    WARNING: For now, TorchSynthModules should be atomic and not contain other
+    SynthModules.
+    TODO: Later, we should deprecate SynthModule and fold everything into here.
+    """
+
+    def __init__(self, sample_rate: int = SAMPLE_RATE):
+        """
+        NOTE:
+        __init__ should only set parameters.
+        We shouldn't be doing computations in __init__ because
+        the computations will change when the parameters change.
+        """
+        nn.Module.__init__(self)
+        self.sample_rate = T(sample_rate)
+        self.modparameters: Dict[ModParameter] = {}
+        self.torchparameters: nn.ParameterDict = nn.ParameterDict()
+
+    def seconds_to_samples(self, seconds: T) -> T:
+        return torch.round(seconds * self.sample_rate).int()
+
+    def add_modparameters(self, modparameters: List[ModParameter]):
+        """
+        Add parameters to this SynthModule's parameters dictionary.
+        (Since there is inheritance, this might happen several times.)
+        """
+        for modparameter in modparameters:
+            assert modparameter.name not in self.modparameters
+            self.modparameters[modparameter.name] = modparameter
+
+            assert modparameter.name not in self.torchparameters
+            # TODO: I'm not 100% sure it's kosher to add nn.Parameters
+            # outside of __init__, but here we go.
+            # TODO: Internally we want to store all torch modparameter
+            # values using their 0/1 range, not their human-interpretable
+            # range.
+            # We might also rethink the syntactic sugar, e.g. sometimes
+            # we want to expose the raw 0/1 and sometimes we want to expose the
+            # clipped one.
+            self.torchparameters[modparameter.name] = nn.Parameter(
+                T(modparameter.value)
+            )
+
+    def forward(self, *inputs: Any) -> T:  # pragma: no cover
+        """
+        Each TorchSynthModule should override this.
+        """
+        pass
+
+    def npyforward(self, *inputs: Any) -> np.ndarray:  # pragma: no cover
+        """
+        This is the numpy version of the torch.nn.Module.forward command.
+        All torch.tensor inputs and outputs are cast to ndarrays.
+        """
+        npyinput = []
+        for i in inputs:
+            if isinstance(i, T):
+                npyinput.append(i.numpy())
+            else:
+                npyinput.append(i)
+        return self.forward(*npyinput).numpy()
+
+    # The following is cheesy AF but needed because the
+    # torchparameter is the master variable.
+    # TODO: Remove .value from modparameter
+    def _update_modparameters(self) -> None:
+        for modparameter_id in self.modparameters:
+            self.modparameters[modparameter_id].set_value(
```

Comment:
This should be set_value_0to1 no? That is the range we are getting back from torch parameters correct ([0,1])?

Reply:
@jorshi no, right now we are wrapping the existing behavior so torchparameters are in the original range. I do think we should change it to 0, 1 internally tho